core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java

```diff
@@ -0,0 +1,445 @@
+            }
+
+            if (!query.hasNext()) {
+                return maybeValue;
+            }
+
+            int next = Byte.toUnsignedInt(query.next());
+            switch (edgeType(metadata)) {
+                case EDGES_SINGLE:
+                    if (next != Byte.toUnsignedInt(nodes.get(movingOffset++))) {
+                        return NOT_FOUND;
+                    }
+                    movingOffset = nodes.getLong(movingOffset);
+                    break;
+                case EDGES_BITSET: {
+                    int min = Byte.toUnsignedInt(nodes.get(movingOffset++));
+                    int max = Byte.toUnsignedInt(nodes.get(movingOffset++));
+                    if (min > next || max < next || !BufferBitSet.get(nodes, movingOffset, next - min)) {
+                        return NOT_FOUND;
+                    }
+                    int index = BufferBitSet.cardinalityTo(nodes, movingOffset, next - min);
+                    movingOffset += BufferBitSet.arraySize(max - min + 1) * Long.BYTES;
+                    movingOffset = nodes.getLong(movingOffset + index * Long.BYTES);
+                    break;
+                }
+                case EDGES_CONDENSED: {
+                    int min = Byte.toUnsignedInt(nodes.get(movingOffset++));
+                    int max = Byte.toUnsignedInt(nodes.get(movingOffset++));
+                    if (min > next || max < next) {
+                        return NOT_FOUND;
+                    }
+                    int index = next - min;
+                    movingOffset = nodes.getLong(movingOffset + index * Long.BYTES);
+                    break;
+                }
+                case EDGES_NONE:
+                    return NOT_FOUND;
+            }
+        }
+    }
+
+    private int takeFirstValueOnLeft(long movingOffset) {
+        while (true) {
+            final int metadata = Byte.toUnsignedInt(nodes.get(movingOffset));
+            movingOffset += Byte.BYTES;
+
+            if (hasInfix(metadata)) { // there is an infix
+                int infixSize = nodes.getInt(movingOffset);
+                movingOffset += Integer.BYTES;
+                movingOffset += infixSize * Byte.BYTES;
+            }
+
+            if (hasValue(metadata)) { // there is a value
+                return nodes.getInt(movingOffset);
+            }
+
+            switch (edgeType(metadata)) {
+                case EDGES_BITSET:
+                    int min = Byte.toUnsignedInt(nodes.get(movingOffset++));
+                    int max = Byte.toUnsignedInt(nodes.get(movingOffset++));
+
+                    movingOffset += BufferBitSet.arraySize(max - min + 1) * Long.BYTES;
+                    movingOffset = nodes.getLong(movingOffset);
+                    break;
+                case EDGES_CONDENSED:
+                    movingOffset += 2 * Byte.BYTES;
+                    movingOffset = nodes.getLong(movingOffset);
+                    break;
+                default: // should be unreachable
```

Comment:
but if it is, maybe NOT_FOUND?

Reply:
There is no correct way to prove this. This case may be unreachable only if a serialized index is corrupted (which may not because we check checksums on the database).

default indicates that I have verified that this case should not happen on regular execution.

Reply:
I'd throw `UnsupportedOperationException` if it is possible to maintain 100% coverage.